File: sidecars/vpcal/src/vpcal/core/framing_match.py
"""Framing-guidance match score (topology / coverage level, not pixel align).

Used by the stills capture window: hit-rate of expected cabinets vs observed,
plus a coarse bbox-area ratio term. Hysteresis avoids green/red flicker.
"""

from __future__ import annotations

import math
from collections.abc import Iterable, Sequence

Cabinet = tuple[int, int]  # (col, row)
BBox = Sequence[float]  # [x0, y0, x1, y1] normalized 0..1

HIT_WEIGHT = 0.75
BBOX_WEIGHT = 0.25
MATCH_ENTER = 80.0
MATCH_EXIT = 70.0


def cabinets_norm_bbox(
    cabinets: Iterable[Cabinet],
    cols: int,
    rows: int,
) -> list[float] | None:
    """Axis-aligned bbox of cabinets in normalized screen UV (0..1)."""
    cols = max(1, int(cols))
    rows = max(1, int(rows))
    items = list(cabinets)
    if not items:
        return None
    c0 = min(c for c, _r in items)
    c1 = max(c for c, _r in items)
    r0 = min(r for _c, r in items)
    r1 = max(r for _c, r in items)
    return [c0 / cols, r0 / rows, (c1 + 1) / cols, (r1 + 1) / rows]


def _area(bbox: BBox | None) -> float:
    if not bbox or len(bbox) < 4:
        return 0.0
    return max(0.0, float(bbox[2]) - float(bbox[0])) * max(
        0.0, float(bbox[3]) - float(bbox[1])
    )


def compute_framing_score(
    expected: Iterable[Cabinet],
    observed: Iterable[Cabinet],
    *,
    expected_bbox: BBox | None = None,
    observed_bbox: BBox | None = None,
) -> float:
    """Return match percent 0..100 (topology hit-rate + bbox area tolerance)."""
    exp = {(int(c), int(r)) for c, r in expected}
    obs = {(int(c), int(r)) for c, r in observed}
    if not exp:
        hit = 1.0 if not obs else 0.0
    else:
        hit = len(exp & obs) / len(exp)

    ea = _area(expected_bbox)
    oa = _area(observed_bbox)
    if ea > 1e-9 and oa > 1e-9:
        ratio = oa / ea
        if 0.6 <= ratio <= 1.5:
            bbox_s = 1.0
        else:
            # Decay outside the tolerance band (log-space, ~3× → 0).
            bbox_s = max(0.0, 1.0 - abs(math.log(ratio)) / math.log(3.0))
    else:
        bbox_s = hit

    return round(100.0 * (HIT_WEIGHT * hit + BBOX_WEIGHT * bbox_s), 2)


def apply_match_hysteresis(
    score: float,
    matched: bool,
    *,
    enter: float = MATCH_ENTER,
    exit_: float = MATCH_EXIT,
) -> bool:
    """Enter green at ``enter``, leave below ``exit_`` (default 80 / 70)."""
    if matched:
        return float(score) >= float(exit_)
    return float(score) >= float(enter)


def missing_cabinets_hint(
    expected: Iterable[Cabinet],
    observed: Iterable[Cabinet],
) -> str:
    """Short Chinese hint listing missing expected cabinets."""
    exp = {(int(c), int(r)) for c, r in expected}
    obs = {(int(c), int(r)) for c, r in observed}
    miss = sorted(exp - obs)
    if not miss:
        return "匹配达标 · 画面稳定即自动拍摄"
    sample = ", ".join(f"{c}×{r}" for c, r in miss[:3])
    more = f" 等 {len(miss)} 个" if len(miss) > 3 else f" · 共 {len(miss)} 个"
    return f"还差箱体 {sample}{more}"


def summarize_detections(dets, image_shape) -> dict:
    """Build detect_state extras from ``detect_markers`` output.

    Returns ``{count, cabinets: [[screen_id,col,row],...], bbox_frac}``.
    """
    from vpcal.core.observations import MarkerId

    cabinets: list[list[int]] = []
    seen: set[tuple[int, int, int]] = set()
    us: list[float] = []
    vs: list[float] = []
    for d in dets:
        mid = d.marker_id
        if isinstance(mid, MarkerId):
            key = (mid.screen_id, mid.cab_col, mid.cab_row)
            if key not in seen:
                seen.add(key)
                cabinets.append([mid.screen_id, mid.cab_col, mid.cab_row])
        us.append(float(d.pixel_u))
        vs.append(float(d.pixel_v))
    h = int(image_shape[0]) if image_shape is not None else 0
    w = int(image_shape[1]) if image_shape is not None and len(image_shape) > 1 else 0
    bbox_frac = None
    if us and w > 0 and h > 0:
        # Quantize so 1px jitter does not defeat detect_state emit dedup.
        bbox_frac = [
            round(max(0.0, min(us) / w), 3),
            round(max(0.0, min(vs) / h), 3),
            round(min(1.0, max(us) / w), 3),
            round(min(1.0, max(vs) / h), 3),
        ]
    return {"count": len(dets), "cabinets": cabinets, "bbox_frac": bbox_frac}

File: src/volo/lib/framingMatch.ts
/** Framing match score/hysteresis mirrors vpcal.core.framing_match; UI copy is frontend-owned. */

export type Cabinet = readonly [number, number];
/** Axis-aligned [x0, y0, x1, y1] normalized 0..1. */
export type BBox = readonly [number, number, number, number];

export const MATCH_HIT_WEIGHT = 0.75;
export const MATCH_BBOX_WEIGHT = 0.25;
export const MATCH_ENTER = 80;
export const MATCH_EXIT = 70;
export const FRAMING_MATCHED_HINT = "匹配达标 · 画面稳定即自动拍摄";

const ZONE_DIFF: Record<string, string> = {
  左半: "还差左侧区域",
  右半: "还差右侧区域",
  顶边: "构图偏低，向上抬镜",
  底边: "构图偏高，向下压镜",
  全景: "再退后一步收全景",
  中央: "还差中央区域",
};

export function cabinetsNormBBox(
  cabinets: Iterable<Cabinet>,
  cols: number,
  rows: number,
): BBox | null {
  const cN = Math.max(1, cols | 0);
  const rN = Math.max(1, rows | 0);
  const items = Array.from(cabinets);
  if (!items.length) return null;
  let c0 = Infinity, c1 = -Infinity, r0 = Infinity, r1 = -Infinity;
  for (const [c, r] of items) {
    if (c < c0) c0 = c;
    if (c > c1) c1 = c;
    if (r < r0) r0 = r;
    if (r > r1) r1 = r;
  }
  return [c0 / cN, r0 / rN, (c1 + 1) / cN, (r1 + 1) / rN];
}

/** [x0,y0,x1,y1] → handoff GuideThumb [x,y,w,h]; null if empty. */
export function bboxToXywh(bbox: BBox | null): readonly [number, number, number, number] | null {
  if (!bbox) return null;
  const w = bbox[2] - bbox[0];
  const h = bbox[3] - bbox[1];
  if (w <= 0 || h <= 0) return null;
  return [bbox[0], bbox[1], w, h];
}

function area(bbox: BBox | null | undefined): number {
  if (!bbox) return 0;
  return Math.max(0, bbox[2] - bbox[0]) * Math.max(0, bbox[3] - bbox[1]);
}

function zoneOfBBox(bbox: BBox): string {
  const cx = (bbox[0] + bbox[2]) / 2;
  const cy = (bbox[1] + bbox[3]) / 2;
  const w = bbox[2] - bbox[0];
  const h = bbox[3] - bbox[1];
  if (w >= 0.7 && h >= 0.7) return "全景";
  if (cy < 0.35) return "顶边";
  if (cy > 0.65) return "底边";
  if (cx < 0.4) return "左半";
  if (cx > 0.6) return "右半";
  return "中央";
}

function roleLabel(role: string): string {
  const map: Record<string, string> = {
    fan: "扇形机位",
    top: "上沿",
    bottom: "下沿",
    added: "补位",
  };
  return map[role] || role || "机位";
}

export function computeFramingScore(
  expected: Iterable<Cabinet>,
  observed: Iterable<Cabinet>,
  expectedBbox?: BBox | null,
  observedBbox?: BBox | null,
): number {
  const exp = new Set(Array.from(expected, ([c, r]) => `${c | 0},${r | 0}`));
  const obs = new Set(Array.from(observed, ([c, r]) => `${c | 0},${r | 0}`));
  let hit = 0;
  if (!exp.size) hit = obs.size ? 0 : 1;
  else {
    let n = 0;
    for (const k of exp) if (obs.has(k)) n += 1;
    hit = n / exp.size;
  }

  const ea = area(expectedBbox);
  const oa = area(observedBbox);
  let bboxS = hit;
  if (ea > 1e-9 && oa > 1e-9) {
    const ratio = oa / ea;
    if (ratio >= 0.6 && ratio <= 1.5) bboxS = 1;
    else bboxS = Math.max(0, 1 - Math.abs(Math.log(ratio)) / Math.log(3));
  }
  return Math.round((100 * (MATCH_HIT_WEIGHT * hit + MATCH_BBOX_WEIGHT * bboxS)) * 100) / 100;
}

export function applyMatchHysteresis(
  score: number,
  matched: boolean,
  enter = MATCH_ENTER,
  exit = MATCH_EXIT,
): boolean {
  return matched ? score >= exit : score >= enter;
}

/** Region chip under guide thumb. Pass `box` when already computed for the station. */
export function stationRegionLabel(
  role: string | undefined,
  covers: Iterable<Cabinet>,
  cols: number,
  rows: number,
  box?: BBox | null,
): string {
  const b = box === undefined ? cabinetsNormBBox(covers, cols, rows) : box;
  const zone = b ? zoneOfBBox(b) : "";
  const base = roleLabel(role || "");
  return zone ? `${base} · ${zone}` : base;
}

/**
 * Unmatched diff copy for match badge + guide side.
 * Empty when expected cabinets are all observed (caller uses FRAMING_MATCHED_HINT).
 */
export function framingDiffHint(
  expected: Iterable<Cabinet>,
  observed: Iterable<Cabinet>,
  cols: number,
  rows: number,
): string {
  const exp = Array.from(expected, ([c, r]) => [c | 0, r | 0] as Cabinet);
  const obs = new Set(Array.from(observed, ([c, r]) => `${c | 0},${r | 0}`));
  const miss = exp.filter(([c, r]) => !obs.has(`${c},${r}`));
  if (!miss.length) return "";
  const missBox = cabinetsNormBBox(miss, cols, rows);
  if (!missBox) return "还差推荐覆盖区域";
  return ZONE_DIFF[zoneOfBBox(missBox)] || "还差推荐覆盖区域";
}
